### PR TITLE
Migrate masked_fill__cuda to ATen

### LINKED ISSUE #49543: Migrate `masked_fill` from TH to ATen (CUDA)
Porting TH operators is essential for code simplicity and performance reasons.

Porting guides and Q&A are available in umbrella issue: #24507

Feel free to add @VitalyFedyunin as a reviewer to get a prioritized review.

### PR DESCRIPTION
**Pull request by kshitij12345**
Fixes #49543

### CODE REVIEW

**Comment by kshitij12345** [comment]
**Benchmark**

```
[---------------------- masked_fill_ mdtype torch.uint8 input dtypetorch.int32 Type: Scalar -----------------------]
                             |  After PR torch.masked_fill_ Scalar CUDA  |  Before PR torch.masked_fill_ Scalar CUDA
1 threads: ---------------------------------------------------------------------------------------------------------
      (128,)                 |                     19.2                  |                     21.0                 
      (256, 128)             |                     19.4                  |                     22.6                 
      (2, 512, 256)          |                     20.2                  |                     21.8                 
      (2, 64, 256, 128)      |                     69.1                  |                     73.5                 
      (4, 2, 512, 256, 128)  |                   2167.2                  |                   2383.0                 

Times are in microseconds (us).

[-------------------------- masked_fill_ mdtype torch.uint8 input dtypetorch.int32 Type: Scalar Tensor --------------------------]
                             |  After PR torch.masked_fill_ Scalar Tensor CUDA  |  Before PR torch.masked_fill_ Scalar Tensor CUDA
1 threads: -----------------------------------------------------------------------------------------------------------------------
      (128,)                 |                        30.6                      |                         33.0                    
      (256, 128)             |                        31.9                      |                         34.4                    
      (2, 512, 256)          |                        32.0                      |                         33.2                    
      (2, 64, 256, 128)      |                        83.3                      |                         86.3                    
      (4, 2, 512, 256, 128)  |                      2181.6                      |                       2397.3                    

Times are in microseconds (us).

[--------------------- masked_fill_ mdtype torch.uint8 input dtypetorch.float32 Type: Scalar ----------------------]
                             |  After PR torch.masked_fill_ Scalar CUDA  |  Before PR torch.masked_fill_ Scalar CUDA
1 threads: ---------------------------------------------------------------------------------------------------------
      (128,)                 |                     19.4                  |                     20.3                 
      (256, 128)             |                     20.9                  |                     24.1                 
      (2, 512, 256)          |                     21.7                  |                     22.6                 
      (2, 64, 256, 128)      |                     69.1                  |                     73.5                 
      (4, 2, 512, 256, 128)  |                   2166.5                  |                   2382.9                 

Times are in microseconds (us).

[------------------------- masked_fill_ mdtype torch.uint8 input dtypetorch.float32 Type: Scalar Tensor -------------------------]
                             |  After PR torch.masked_fill_ Scalar Tensor CUDA  |  Before PR torch.masked_fill_ Scalar Tensor CUDA
1 threads: -----------------------------------------------------------------------------------------------------------------------
      (128,)                 |                        31.1                      |                         31.5                    
      (256, 128)             |                        33.8                      |                         35.8                    
      (2, 512, 256)          |                        35.1                      |                         34.2                    
      (2, 64, 256, 128)      |                        83.3                      |                         86.7                    
      (4, 2, 512, 256, 128)  |                      2181.2                      |                       2397.6                    

Times are in microseconds (us).

[--------------------- masked_fill_ mdtype torch.uint8 input dtypetorch.float64 Type: Scalar ----------------------]
                             |  After PR torch.masked_fill_ Scalar CUDA  |  Before PR torch.masked_fill_ Scalar CUDA
1 threads: ---------------------------------------------------------------------------------------------------------
      (128,)                 |                     21.1                  |                     24.2                 
      (256, 128)             |                     20.8                  |                     22.5                 
      (2, 512, 256)          |                     20.3                  |                     22.5                 
      (2, 64, 256, 128)      |                    129.4                  |                    136.0                 
      (4, 2, 512, 256, 128)  |                   4099.5                  |                   4327.9                 

Times are in microseconds (us).

[------------------------- masked_fill_ mdtype torch.uint8 input dtypetorch.float64 Type: Scalar Tensor -------------------------]
                             |  After PR torch.masked_fill_ Scalar Tensor CUDA  |  Before PR torch.masked_fill_ Scalar Tensor CUDA
1 threads: -----------------------------------------------------------------------------------------------------------------------
      (128,)                 |                        32.4                      |                         35.9                    
      (256, 128)             |                        32.3                      |                         33.9                    
      (2, 512, 256)          |                        31.9                      |                         33.8                    
      (2, 64, 256, 128)      |                       142.7                      |                        150.6                    
      (4, 2, 512, 256, 128)  |                      4114.7                      |                       4343.5                    

Times are in microseconds (us).

[----------------------- masked_fill_ mdtype torch.bool input dtypetorch.int32 Type: Scalar -----------------------]
                             |  After PR torch.masked_fill_ Scalar CUDA  |  Before PR torch.masked_fill_ Scalar CUDA
1 threads: ---------------------------------------------------------------------------------------------------------
      (128,)                 |                      4.4                  |                      4.4                 
      (256, 128)             |                      4.4                  |                      4.6                 
      (2, 512, 256)          |                      4.3                  |                      4.8                 
      (2, 64, 256, 128)      |                     69.2                  |                     73.2                 
      (4, 2, 512, 256, 128)  |                   2165.0                  |                   2382.1                 

Times are in microseconds (us).

[-------------------------- masked_fill_ mdtype torch.bool input dtypetorch.int32 Type: Scalar Tensor ---------------------------]
                             |  After PR torch.masked_fill_ Scalar Tensor CUDA  |  Before PR torch.masked_fill_ Scalar Tensor CUDA
1 threads: -----------------------------------------------------------------------------------------------------------------------
      (128,)                 |                        12.4                      |                         12.2                    
      (256, 128)             |                        12.3                      |                         12.4                    
      (2, 512, 256)          |                        12.4                      |                         12.7                    
      (2, 64, 256, 128)      |                        81.0                      |                         84.9                    
      (4, 2, 512, 256, 128)  |                      2178.0                      |                       2395.5                    

Times are in microseconds (us).

[---------------------- masked_fill_ mdtype torch.bool input dtypetorch.float32 Type: Scalar ----------------------]
                             |  After PR torch.masked_fill_ Scalar CUDA  |  Before PR torch.masked_fill_ Scalar CUDA
1 threads: ---------------------------------------------------------------------------------------------------------
      (128,)                 |                      4.4                  |                      4.5                 
      (256, 128)             |                      4.4                  |                      4.7                 
      (2, 512, 256)          |                      4.3                  |                      5.0                 
      (2, 64, 256, 128)      |                     69.1                  |                     73.7                 
      (4, 2, 512, 256, 128)  |                   2165.1                  |                   2381.4                 

Times are in microseconds (us).

[------------------------- masked_fill_ mdtype torch.bool input dtypetorch.float32 Type: Scalar Tensor --------------------------]
                             |  After PR torch.masked_fill_ Scalar Tensor CUDA  |  Before PR torch.masked_fill_ Scalar Tensor CUDA
1 threads: -----------------------------------------------------------------------------------------------------------------------
      (128,)                 |                        12.4                      |                         12.3                    
      (256, 128)             |                        12.3                      |                         12.6                    
      (2, 512, 256)          |                        12.5                      |                         12.9                    
      (2, 64, 256, 128)      |                        81.5                      |                         86.1                    
      (4, 2, 512, 256, 128)  |                      2177.9                      |                       2395.1                    

Times are in microseconds (us).

[---------------------- masked_fill_ mdtype torch.bool input dtypetorch.float64 Type: Scalar ----------------------]
                             |  After PR torch.masked_fill_ Scalar CUDA  |  Before PR torch.masked_fill_ Scalar CUDA
1 threads: ---------------------------------------------------------------------------------------------------------
      (128,)                 |                      4.4                  |                      4.5                 
      (256, 128)             |                      4.4                  |                      4.7                 
      (2, 512, 256)          |                      4.3                  |                      4.9                 
      (2, 64, 256, 128)      |                    129.5                  |                    136.1                 
      (4, 2, 512, 256, 128)  |                   4099.5                  |                   4328.4                 

Times are in microseconds (us).

[------------------------- masked_fill_ mdtype torch.bool input dtypetorch.float64 Type: Scalar Tensor --------------------------]
                             |  After PR torch.masked_fill_ Scalar Tensor CUDA  |  Before PR torch.masked_fill_ Scalar Tensor CUDA
1 threads: -----------------------------------------------------------------------------------------------------------------------
      (128,)                 |                        12.4                      |                         12.4                    
      (256, 128)             |                        12.3                      |                         12.4                    
      (2, 512, 256)          |                        12.5                      |                         12.7                    
      (2, 64, 256, 128)      |                       141.9                      |                        148.6                    
      (4, 2, 512, 256, 128)  |                      4112.8                      |                       4342.3                    

Times are in microseconds (us).

```

<details>

<summary> Benchmark Code </summary>

**Benchmark Script**
```python
import torch
import itertools
import time
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import sys
import pickle

print('Using pytorch %s' % (torch.__version__))

shapes = [(128,), (256, 128), (2, 512, 256), (2, 64, 256, 128), (4, 2, 512, 256, 128)]
results = []
repeats = 10

for m_dtype in [torch.uint8, torch.bool]:
    for dtype in [torch.int32, torch.float, torch.double]:
        for mat1_shape in shapes:
            mat1 = torch.zeros(*mat1_shape, dtype=dtype, device='cuda')
            mask = (torch.randn(mat1.shape, device='cuda') > 0).to(m_dtype)
            source = torch.ones(mask.sum(), dtype=dtype, device='cuda')
            scalar_t = torch.tensor(2, device='cuda')
            scalar = 2
            tasks = [("mat1.masked_fill_(mask, scalar)", "After PR torch.masked_fill_ Scalar CUDA", "Scalar"),
                     ("mat1.masked_fill_(mask, scalar_t)", "After PR torch.masked_fill_ Scalar Tensor CUDA", "Scalar Tensor")]
            timers = [Timer(stmt=stmt, label=f"masked_fill_ mdtype {m_dtype} input dtype{dtype} Type: {label}", sub_label=f"{(mat1_shape)}", description=desc, globals=globals()) for stmt, desc, label in tasks]

            for i, timer in enumerate(timers * repeats):
                results.append(
                    timer.blocked_autorange()
                )
                print(f"\r{i + 1} / {len(timers) * repeats}", end="")
                sys.stdout.flush()

with open('after_pr.pkl', 'wb') as f:
    pickle.dump(results, f)

comparison = Compare(results)
comparison.print()
```

**Comparison Script**
```python
import torch
import itertools
import time
from torch.utils.benchmark import Timer
from torch.utils.benchmark import Compare
import sys
import pickle

with open('after_pr.pkl', 'rb') as f:
    after_results = pickle.load(f)

with open('before_pr.pkl', 'rb') as f:
    before_results = pickle.load(f)

comparison = Compare(after_results + before_results)
comparison.print()
```
</details>

**Comment by kshitij12345** [comment]
@ngimel Gentle Ping :)

**Comment by VitalyFedyunin** [comment]
I will review it this Wednesday

**Comment by kshitij12345** [comment]
@VitalyFedyunin Great! Thanks for the update.

File: test/test_torch.py [hunk 1]
```diff
@@ -4600,6 +4570,36 @@ def test_masked_select_discontiguous(self, device):
                 torch.masked_select(v, m, out=out_dc)
                 self.assertEqual(out_dc, expected, atol=0, rtol=0)
 
+    @dtypes(*torch.testing.get_all_dtypes())
```
Comment (ngimel) [inline test/test_torch.py:4573]: you can get dtypes pairs using the syntax
```
 @dtypes(*product(torch.testing.get_all_dtypes(), (torch.uint8, torch.bool)))
 ```
 they are accessible inside the test via dtypes[0], dtypes[1]

File: test/test_torch.py [hunk 2]
```diff
@@ -4600,6 +4570,36 @@ def test_masked_select_discontiguous(self, device):
...
+    def test_masked_fill(self, device, dtype):
+        dt = dtype
+        with warnings.catch_warnings(record=True) as w:
+            warnings.simplefilter("always")
+            for dtype in [torch.uint8, torch.bool]:
+                num_dest = 10
+                dst = torch.tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=dt)
```
Comment (ngimel) [inline test/test_torch.py:4580]: torch.zeros

File: test/test_torch.py [hunk 3]
```diff
@@ -4600,6 +4570,36 @@ def test_masked_select_discontiguous(self, device):
...
+                mask = torch.rand(num_dest).mul(2).floor().to(dtype)
```
Comment (ngimel) [inline test/test_torch.py:4581]: this is a complicated way of saying `torch.randint(2, (num_dest,), dtype=dtype[1])` (assuming your mask dtype is `dtype[1]`

File: test/test_torch.py [hunk 4]
```diff
@@ -4600,6 +4570,36 @@ def test_masked_select_discontiguous(self, device):
...
+
+                dst.masked_fill_(mask, val)
+                for i in range(num_dest):
+                    if mask[i]:
+                        dst2[i] = val
+                self.assertEqual(dst, dst2, atol=0, rtol=0)
+
+                # test non-contiguous case
+                dst = torch.randn(num_dest, num_dest, num_dest).permute((2, 0, 1))
+                dst2 = dst.clone()
```
Comment (ngimel) [inline test/test_torch.py:4593]: since clone by default will preserve layout, this test is not testing much - it's calling similar operations on similar tensors

**Review by ngimel — COMMENTED** [review]
The port itself looks good (thanks for cleaning up THC files!), I left comments about the test.

**Review by ngimel — APPROVED** [review]
